Cannot instatiate view with type parameter (again)

### Description of the bug

Seems like [this bug](https://github.com/vaadin/flow/issues/8371) has reappeared.

When my route has a generic type it results in this error on startup.  Removal of the HasUrlParameter interface, or the type T parameter, will eliminate the issue.

We use version 24.6.6
```
Caused by: java.lang.IllegalArgumentException: Parameter type of the given navigationTarget 'com.idexx.medicalaffairs.admin.web.tests.antibiotic.AntibioticPage' could not be resolved.
at com.vaadin.flow.router.ParameterDeserializer.getClassType(ParameterDeserializer.java:186)
at com.vaadin.flow.router.internal.HasUrlParameterFormat.getTemplate(HasUrlParameterFormat.java:85)
```


### Expected behavior

I should be able to use a generic type in my route

### Minimal reproducible example


```
@PageTitle("Hello World")
@Route("")
public class HelloWorldView<T> extends HorizontalLayout implements HasUrlParameter<String> {
public HelloWorldView() {
add(new Span("Hello"));
}

private String urlParameter;
@Override
public void setParameter(BeforeEvent beforeEvent, @WildcardParameter String parameter) {
urlParameter = parameter;
}
}
```

### Versions

- Vaadin / Flow version:
- Java version:
- OS version:
- Browser version (if applicable):
- Application Server (if applicable):
- IDE (if applicable):

--- a/flow-server/src/main/java/com/vaadin/flow/router/ParameterDeserializer.java
+++ b/flow-server/src/main/java/com/vaadin/flow/router/ParameterDeserializer.java
@@ -180,8 +180,7 @@ public final class ParameterDeserializer {
      * @return parameter type class
      */
     public static Class<?> getClassType(Class<?> navigationTarget) {
-        Type type = GenericTypeReflector.getTypeParameter(navigationTarget,
-                HasUrlParameter.class.getTypeParameters()[0]);
+        Type type = findParameterType(navigationTarget);
         if (!(type instanceof Class<?>)) {
             throw new IllegalArgumentException(String.format(
                     "Parameter type of the given navigationTarget '%s' could not be resolved.",
@@ -209,6 +208,18 @@ public final class ParameterDeserializer {
                 .getFunctionalMethod(HasUrlParameter.class).getName());
 
         // Raw method has no parameter annotations if compiled by Eclipse
+        Type parameterType = findParameterType(navigationTarget);
+        Class<?> parameterClass = GenericTypeReflector.erase(parameterType);
+
+        return Stream.of(navigationTarget.getMethods())
+                .filter(method -> methodName.equals(method.getName()))
+                .filter(method -> hasValidParameterTypes(method,
+                        parameterClass))
+                .anyMatch(method -> method.getParameters()[1]
+                        .isAnnotationPresent(parameterAnnotation));
+    }
+
+    private static Type findParameterType(Class<?> navigationTarget) {
         Type parameterType = GenericTypeReflector.getTypeParameter(
                 navigationTarget, HasUrlParameter.class.getTypeParameters()[0]);
         if (parameterType == null) {
@@ -225,14 +236,7 @@ public final class ParameterDeserializer {
                 }
             }
         }
-        Class<?> parameterClass = GenericTypeReflector.erase(parameterType);
-
-        return Stream.of(navigationTarget.getMethods())
-                .filter(method -> methodName.equals(method.getName()))
-                .filter(method -> hasValidParameterTypes(method,
-                        parameterClass))
-                .anyMatch(method -> method.getParameters()[1]
-                        .isAnnotationPresent(parameterAnnotation));
+        return parameterType;
     }
 
     private static boolean hasValidParameterTypes(Method method,

--- a/flow-server/src/test/java/com/vaadin/flow/router/ParameterDeserializerTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/router/ParameterDeserializerTest.java
@@ -17,6 +17,7 @@ package com.vaadin.flow.router;
 
 import org.junit.Test;
 
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 
@@ -61,6 +62,50 @@ public class ParameterDeserializerTest {
                 OptionalParameter.class));
         assertTrue(ParameterDeserializer.isAnnotatedParameter(
                 GenericAnnotated.class, OptionalParameter.class));
+    }
+
+    @Test
+    public void getClassType_concreteClass_parameterFromInterface() {
+        Class<?> type = ParameterDeserializer.getClassType(Simple.class);
+        assertEquals(String.class, type);
+        type = ParameterDeserializer.getClassType(SimpleAnnotated.class);
+        assertEquals(String.class, type);
+    }
+
+    @Test
+    public void getClassType_concreteClass_parameterFromExtendedInterface() {
+        Class<?> type = ParameterDeserializer.getClassType(Normal.class);
+        assertEquals(String.class, type);
+        type = ParameterDeserializer.getClassType(NormalAnnotated.class);
+        assertEquals(String.class, type);
+    }
+
+    @Test
+    public void getClassType_concreteClass_parameterFromSuperclass() {
+        Class<?> type = ParameterDeserializer
+                .getClassType(ParameterizedViaSuperClass.class);
+        assertEquals(String.class, type);
+        type = ParameterDeserializer
+                .getClassType(ParameterizedAnnotatedClass.class);
+        assertEquals(String.class, type);
+    }
+
+    @Test
+    public void getClassType_parameterizedClass_parameterFromInterface() {
+        Class<?> type = ParameterDeserializer
+                .getClassType(ParameterizedClass.class);
+        assertEquals(String.class, type);
+        type = ParameterDeserializer
+                .getClassType(ParameterizedAnnotatedClass.class);
+        assertEquals(String.class, type);
+    }
+
+    @Test
+    public void getClassType_parameterizedClass_parameterFromParameterizedInterface() {
+        Class<?> type = ParameterDeserializer.getClassType(Generic.class);
+        assertEquals(String.class, type);
+        type = ParameterDeserializer.getClassType(GenericAnnotated.class);
+        assertEquals(String.class, type);
     }
 
     public static class Simple implements HasUrlParameter<String> {

